src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+      return false;
+    }
+    return true;
+  }
+  LogicalResult matchAndRewrite(ONNXGemmOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+    int64_t transA = adaptor.transA();
+    int64_t transB = adaptor.transB();
+    FloatAttr alpha = adaptor.alphaAttr();
+    FloatAttr beta = adaptor.betaAttr();
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+    auto shapeA = A.getType().dyn_cast<ShapedType>().getShape();
+    auto shapeB = B.getType().dyn_cast<ShapedType>().getShape();
+    auto resultType = getTypeConverter()->convertType(op.getResult().getType()).cast<TensorType>();
+    
+    // C is optional, if it's not there, we need to be aware of it for later computations
+    bool isCPresent = C.getType().isa<TensorType>();
+    // If it's not present, ONNX creates a NoValue operator. Remove it.
+    if (!isCPresent) {
+      rewriter.eraseOp(C.getDefiningOp());
+    }
+
+    // If legal, create a FullyConnected operator instead
+    if (checkLegalTosaFCOp(op, adaptor)) {
+      return rewriteToTosaFC(op, adaptor, rewriter);
+    }
+    
+    // ONNX gives 2d matrix as input and expect a 2d output. TOSA expects everything to be 3D. As such, there
+    // is a need to add reshapes operators before and after we do any computation. We set the batch as 1 as it
+    // is unknown.
+    llvm::SmallVector<int64_t> newShapeA{1};
+    llvm::SmallVector<int64_t> newShapeB{1};
+    newShapeA.append({shapeA[0], shapeA[1]});
+    newShapeB.append({shapeB[0], shapeB[1]});
+
+    A = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(AType.getElementType()), A,
+        rewriter.getI64ArrayAttr(newShapeA)).getResult();
+    B = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(BType.getElementType()), B,
+        rewriter.getI64ArrayAttr(newShapeB)).getResult();
+
+    auto tosaResult = UnrankedTensorType::get(resultType.getElementType());
+
+    // If transA or transB are present, create Transpose operators.
+    if (transA) {
+      Value targetTensor = mlir::tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 1}, {3}).value();
+      Type outputType = UnrankedTensorType::get(AType.getElementType());
+      A = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, A, targetTensor).getResult();
+    }
+    if (transB) {
+      Value targetTensor = mlir::tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 1}, {3}).value();
+      Type outputType = UnrankedTensorType::get(AType.getElementType());
+      B = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, B, targetTensor).getResult();
+    }
+    
+    Value alphaMulResult = NULL;
+    Value betaMulResult = NULL;
```

Comment:
If you initialized these to `A` and `C` respectively, the logic further down would be simplified.